### ISSUE TITLE
Fix S1313: Exclude local IPv4-mapped IPv6 address

### ISSUE THREAD

**Issue by hendrik-buchwald-sonarsource** [SonarSource/sonar-dotnet#6080]
S1313 should not report an issue if a local IPv4-mapped IPv6 address is used.

This means, no issue should be raised for IP address between `::ffff:0:127.0.0.1` to `::ffff:0:127.255.255.255` and `::ffff:127.0.0.1` to `::ffff:127.255.255.255` (starts with `::ffff:0:127.` or `::ffff:127.`). Ignore the casing.

**Comment by martin-strecker-sonarsource** [comment]
Hello @hendrik-buchwald-sonarsource 

I started the implementation and have the following questions regarding the loopback specification. According to [rfc4291 *2.5.5 IPv6 Addresses with Embedded IPv4 Addresses*](https://www.rfc-editor.org/rfc/rfc4291.html#section-2.5.5) there are two different kinds of embeddings:

* IPv4-Compatible IPv6 Address `::127.0.0.1` [deprecated]
* IPv4-Mapped IPv6 Address `::FFFF:127.0.0.1`

The implementation of [`IPAdress`](https://learn.microsoft.com/dotnet/api/system.net.ipaddress) in .NET only supports the last form:

| Form | Example | IsIPv4MappedToIPv6 | IsLoopback | IsLoopback(ipAddress.MapToIPv4()) |
| --- | --- | --- | --- | --- |
| IPv4-Compatible IPv6 Address | `::127.0.0.1` | `false` | `false` | `true` |
| IPv4-Mapped | `::FFFF:127.0.0.1` | `true` | `false` | `true` |
| Not IPv4 mapped | `2001:0db8::127.0.0.1` | `false` | `false` | `true` |
| IP6 loopback | `::1` | `false` | `true` | `false` |

As you can see `MapToIPv4` always succeeds even though `IsIPv4MappedToIPv6` returns false (Just uses the last 4 bytes of the IP6 to return an IP4). `IsIPv4MappedToIPv6` also returns false for *IPv4-Compatible IPv6 Address* which is right according to the spec.
I also haven't found anything regarding the `::ffff:0:127.0.0.1` form. Do you have any sources for it? I would propose to support the following two kinds of loopback addresses for IP6:

* `::1` https://www.rfc-editor.org/rfc/rfc4291.html#section-2.5.3
* `FFFF:127.0.0.1` IPv4-Mapped loopback addresses https://www.rfc-editor.org/rfc/rfc4291.html#section-2.5.5.2

What do you think?

**Comment by hendrik-buchwald-sonarsource** [comment]
Hey @martin-strecker-sonarsource ,

In regards to the`::ffff:0:`, it is related to IPv6 translation. I am not sure in which RFC it is specified. It is the original false-positive that was [reported](https://discuss.sonarsource.com/t/s1313-fp-on-local-ipv4-mapped-ipv6-address/10996). The syntax can also be found [here](https://en.wikipedia.org/wiki/IPv6_address#Special_addresses).

If we do not raise an issue at all for this notation that is fine as well and then I would just proceed as you suggested.

**Comment by martin-strecker-sonarsource** [comment]
I found the original RFC for IPv6 translation. It is [rfc2765](https://www.rfc-editor.org/rfc/rfc2765.html#section-2.1) which was obsoleted (twice). None of the follow-up RFCs mention this form anymore and IPv6 translation is defined differently now ([prefix `64:ff9b::`](https://www.rfc-editor.org/rfc/rfc6052.html#section-2.1)).
A nice tool to find the range name and the RFC is http://www.gestioip.net/cgi-bin/subnet_calculator.cgi in case you are interested (that way I found rfc2765).
For simplicity's sake, I will stick with the current implementation `ipAddress.IsIPv4MappedToIPv6 && IPAddress.IsLoopback(ipAddress.MapToIPv4())` which only takes mapped IP4 addressed into account and none of the translated forms.
